File: backend/app/schemas/project.py
"""Schemas Pydantic cho Enterprise Project Management."""
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.project import ProjectMemberRole, ProjectPriority, ProjectStatus


# ═══════════════════════════════════════════════════════════════════════════
# Project CRUD
# ═══════════════════════════════════════════════════════════════════════════

class ProjectCreate(BaseModel):
    name: str             = Field(..., min_length=1, max_length=255)
    code: str | None      = Field(default=None, max_length=50)
    description: str | None = Field(default=None, max_length=1000)
    status: ProjectStatus = ProjectStatus.PLANNING
    priority: ProjectPriority = ProjectPriority.MEDIUM
    department_id: int | None = None
    manager_id: int | None    = None
    start_date: date | None   = None
    end_date: date | None     = None
    estimated_hours: float | None  = None
    estimated_budget: float | None = None
    project_weight: float = Field(default=1.0, ge=0.1, le=10)


class ProjectUpdate(BaseModel):
    name: str | None          = Field(default=None, min_length=1, max_length=255)
    code: str | None          = Field(default=None, max_length=50)
    description: str | None   = Field(default=None, max_length=1000)
    status: ProjectStatus | None    = None
    priority: ProjectPriority | None = None
    department_id: int | None = None
    manager_id: int | None    = None
    start_date: date | None   = None
    end_date: date | None     = None
    estimated_hours: float | None  = None
    estimated_budget: float | None = None
    project_weight: float | None = Field(default=None, ge=0.1, le=10)
    reason: str | None        = None   # lý do đổi status (ghi vào history)


class ProjectMemberRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    user_id: int
    full_name: str = ""
    email: str     = ""
    role: ProjectMemberRole
    contribution_share: float = 0
    is_active: bool = True
    joined_at: datetime

    @classmethod
    def from_member(cls, m: Any) -> "ProjectMemberRead":
        return cls(
            id=m.id, user_id=m.user_id,
            full_name=m.user.full_name if m.user else "",
            email=m.user.email if m.user else "",
            role=m.role,
            contribution_share=m.contribution_share or 0,
            is_active=bool(m.is_active),
            joined_at=m.joined_at,
        )


class AssignableUserRead(BaseModel):
    id: int
    full_name: str
    email: str
    department_id: int
    project_role: ProjectMemberRole


class MilestoneRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    title: str
    description: str | None
    due_date: date | None
    is_completed: bool
    completed_at: datetime | None
    weight: int


class MilestoneCreate(BaseModel):
    title: str        = Field(..., min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=1000)
    due_date: date | None   = None
    weight: int             = Field(default=1, ge=1, le=10)


class MilestoneUpdate(BaseModel):
    title: str | None       = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=1000)
    due_date: date | None   = None
    weight: int | None      = Field(default=None, ge=1, le=10)


class StatusHistoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    from_status: str | None
    to_status: str
    actor_name: str = ""
    reason: str | None
    changed_at: datetime

    @classmethod
    def from_history(cls, h: Any) -> "StatusHistoryRead":
        return cls(
            id=h.id,
            from_status=h.from_status,
            to_status=h.to_status,
            actor_name=h.actor.full_name if h.actor else "System",
            reason=h.reason,
            changed_at=h.changed_at,
        )


class AuditLogRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    field_name: str
    old_value: str | None
    new_value: str | None
    actor_name: str = ""
    changed_at: datetime

    @classmethod
    def from_audit(cls, a: Any) -> "AuditLogRead":
        return cls(
            id=a.id, field_name=a.field_name,
            old_value=a.old_value, new_value=a.new_value,
            actor_name=a.actor.full_name if a.actor else "System",
            changed_at=a.changed_at,
        )


# ── Response tổng hợp ──────────────────────────────────────────────────────

class ProjectListItem(BaseModel):
    """Response nhẹ cho danh sách project."""
    model_config = ConfigDict(from_attributes=True)
    id: int
    name: str
    description: str | None
    code: str | None
    status: str
    priority: str
    progress_percentage: float
    start_date: date | None
    end_date: date | None
    estimated_hours: float | None = None
    estimated_budget: float | None = None
    department_id: int | None = None
    manager_id: int | None    = None
    department_name: str = ""
    manager_name: str    = ""
    total_tasks: int     = 0
    completed_tasks: int = 0
    done_tasks: int      = 0
    task_completion_percentage: float = 0
    project_progress_percentage: float = 0
    overdue_tasks: int   = 0
    member_count: int    = 0
    total_members: int   = 0
    milestone_count: int = 0
    milestones_done: int = 0
    is_overdue: bool     = False
    project_weight: float = 1


class TaskSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    title: str
    status: str
    priority: str
    deadline: date | None
    done_at: datetime | None = None
    base_weight: int = 1
    assignee_id: int | None = None
    assignee_name: str = ""
    project_id: int | None = None
    department_id: int | None = None
    is_overdue: bool   = False


class ProjectKpiContribution(BaseModel):
    """KPI breakdown cho project."""
    total_score_contributed: float = 0
    tasks_completed: int           = 0
    tasks_overdue: int             = 0
    milestones_completed: int      = 0
    on_time_rate: float            = 0
    member_contributions: list[dict] = []


class ProjectAnalytics(BaseModel):
    """Analytics đầy đủ cho 1 project."""
    progress_percentage: float
    project_progress_percentage: float = 0
    total_tasks: int
    completed_tasks: int
    done_tasks: int = 0
    pending_tasks: int
    todo_tasks: int = 0
    doing_tasks: int
    review_tasks: int = 0
    blocked_tasks: int
    overdue_tasks: int
    completion_rate: float
    task_completion_percentage: float = 0
    on_time_rate: float
    velocity: float               # tasks/tuần trung bình
    estimated_hours: float | None
    actual_hours: float
    budget_utilization: float | None
    milestone_progress: float     # % milestones hoàn thành
    total_members: int = 0
    total_milestones: int = 0
    completed_milestones: int = 0
    milestone_completion_percentage: float = 0
    risk_level: str               # LOW / MEDIUM / HIGH / CRITICAL
    risk_indicators: list[str]


class ProjectMemberPerformanceRead(BaseModel):
    user_id: int
    full_name: str
    email: str
    department_name: str = ""
    project_role: ProjectMemberRole
    contribution_share: float = 0
    total_tasks: int = 0
    done_tasks: int = 0
    overdue_tasks: int = 0
    task_completion_percentage: float = 0
    kpi_score: float = 0


class ProjectReportRead(BaseModel):
    analytics: ProjectAnalytics
    task_status_breakdown: dict[str, int]
    member_performance: list[ProjectMemberPerformanceRead] = []
    top_contributor: ProjectMemberPerformanceRead | None = None
    most_overdue_member: ProjectMemberPerformanceRead | None = None


class ProjectOverview(BaseModel):
    """Response đầy đủ cho Project Detail page."""
    # Thông tin cơ bản
    id: int
    name: str
    code: str | None
    description: str | None
    status: str
    priority: str
    progress_percentage: float
    start_date: date | None
    end_date: date | None
    estimated_hours: float | None
    actual_hours: float
    estimated_budget: float | None
    project_weight: float = 1
    department_id: int | None = None
    manager_id: int | None = None
    department_name: str = ""
    manager_name: str    = ""
    created_at: datetime | None
    updated_at: datetime | None

    # Liên quan
    members: list[ProjectMemberRead]        = []
    milestones: list[MilestoneRead]         = []
    recent_tasks: list[TaskSummary]         = []
    status_history: list[StatusHistoryRead] = []
    recent_audit_logs: list[AuditLogRead]   = []
    analytics: ProjectAnalytics | None      = None
    kpi_contribution: ProjectKpiContribution | None = None


# ── Member management ──────────────────────────────────────────────────────

class AddMemberRequest(BaseModel):
    user_id: int
    role: ProjectMemberRole = ProjectMemberRole.MEMBER
    contribution_share: float = Field(default=0, ge=0, le=100)
    is_active: bool = True


class UpdateMemberRoleRequest(BaseModel):
    role: ProjectMemberRole | None = None
    contribution_share: float | None = Field(default=None, ge=0, le=100)
    is_active: bool | None = None


class MyProjectRead(BaseModel):
    project_id: int
    project_code: str | None
    project_name: str
    description: str | None = None
    department: str = ""
    project_status: str
    project_role: ProjectMemberRole | None = None
    contribution_share: float = 0
    start_date: date | None = None
    due_date: date | None = None
    progress: float = 0
    project_health: str = "OK"
    assigned_tasks: int = 0
    doing_tasks: int = 0
    review_tasks: int = 0
    done_tasks: int = 0
    overdue_tasks: int = 0


class TeamWorkloadRead(BaseModel):
    user_id: int
    full_name: str
    email: str
    active_projects: int = 0
    assigned_tasks: int = 0
    doing_tasks: int = 0
    review_tasks: int = 0
    overdue_tasks: int = 0
    estimated_hours: float = 0
    actual_hours: float = 0
    workload_status: str = "NORMAL"

File: frontend/src/types/project.ts
// ── Types ──────────────────────────────────────────────────────────────────

export type ProjectStatus =
  | 'PLANNING' | 'ACTIVE' | 'PAUSED' | 'ON_HOLD' | 'REVIEW'
  | 'COMPLETED' | 'CANCELLED' | 'ARCHIVED';

export type ProjectPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
export type MemberRole = 'PROJECT_MANAGER' | 'TEAM_LEAD' | 'MEMBER' | 'VIEWER';

export interface ProjectListItem {
  id: number;
  name: string;
  description?: string | null;
  code: string | null;
  status: ProjectStatus;
  priority: ProjectPriority;
  progress_percentage: number;
  start_date: string | null;
  end_date: string | null;
  estimated_hours: number | null;
  estimated_budget: number | null;
  department_id: number | null;
  manager_id: number | null;
  department_name: string;
  manager_name: string;
  total_tasks: number;
  completed_tasks: number;
  done_tasks: number;
  task_completion_percentage: number;
  project_progress_percentage: number;
  overdue_tasks: number;
  member_count: number;
  total_members: number;
  milestone_count: number;
  milestones_done: number;
  is_overdue: boolean;
  project_weight: number;
}

export interface ProjectMember {
  id: number;
  user_id: number;
  full_name: string;
  email: string;
  role: MemberRole;
  contribution_share: number;
  is_active: boolean;
  joined_at: string;
}

export interface Milestone {
  id: number;
  title: string;
  description: string | null;
  due_date: string | null;
  is_completed: boolean;
  completed_at: string | null;
  weight: number;
}

export interface TaskSummary {
  id: number;
  title: string;
  status: string;
  priority: string;
  deadline: string | null;
  done_at: string | null;
  base_weight: number;
  assignee_id: number | null;
  assignee_name: string;
  project_id: number | null;
  department_id: number | null;
  is_overdue: boolean;
}

export interface AssignableUser {
  id: number;
  full_name: string;
  email: string;
  department_id: number;
  project_role: MemberRole;
}

export interface StatusHistoryItem {
  id: number;
  from_status: string | null;
  to_status: string;
  actor_name: string;
  reason: string | null;
  changed_at: string;
}

export interface AuditLogItem {
  id: number;
  field_name: string;
  old_value: string | null;
  new_value: string | null;
  actor_name: string;
  changed_at: string;
}

export interface ProjectAnalytics {
  progress_percentage: number;
  project_progress_percentage: number;
  total_tasks: number;
  completed_tasks: number;
  pending_tasks: number;
  doing_tasks: number;
  blocked_tasks: number;
  review_tasks: number;
  overdue_tasks: number;
  completion_rate: number;
  task_completion_percentage: number;
  done_tasks: number;
  todo_tasks: number;
  on_time_rate: number;
  velocity: number;
  estimated_hours: number | null;
  actual_hours: number;
  budget_utilization: number | null;
  milestone_progress: number;
  total_members: number;
  total_milestones: number;
  completed_milestones: number;
  milestone_completion_percentage: number;
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  risk_indicators: string[];
}

export interface ProjectMemberPerformance {
  user_id: number;
  full_name: string;
  email: string;
  department_name: string;
  project_role: MemberRole;
  contribution_share: number;
  total_tasks: number;
  done_tasks: number;
  overdue_tasks: number;
  task_completion_percentage: number;
  kpi_score: number;
}

export interface ProjectReport {
  analytics: ProjectAnalytics;
  task_status_breakdown: Record<string, number>;
  member_performance: ProjectMemberPerformance[];
  top_contributor: ProjectMemberPerformance | null;
  most_overdue_member: ProjectMemberPerformance | null;
}

export interface KpiContribution {
  total_score_contributed: number;
  tasks_completed: number;
  tasks_overdue: number;
  milestones_completed: number;
  on_time_rate: number;
  member_contributions: Array<{
    user_id: number;
    full_name: string;
    role: string;
    kpi_score: number;
  }>;
}

export interface ProjectOverview {
  id: number;
  name: string;
  code: string | null;
  description: string | null;
  status: ProjectStatus;
  priority: ProjectPriority;
  progress_percentage: number;
  start_date: string | null;
  end_date: string | null;
  estimated_hours: number | null;
  actual_hours: number;
  estimated_budget: number | null;
  project_weight: number;
  department_id: number | null;
  manager_id: number | null;
  department_name: string;
  manager_name: string;
  created_at: string | null;
  updated_at: string | null;
  members: ProjectMember[];
  milestones: Milestone[];
  recent_tasks: TaskSummary[];
  status_history: StatusHistoryItem[];
  recent_audit_logs: AuditLogItem[];
  analytics: ProjectAnalytics | null;
  kpi_contribution: KpiContribution | null;
}

export interface ProjectCreate {
  name: string;
  code?: string;
  description?: string;
  status?: ProjectStatus;
  priority?: ProjectPriority;
  department_id?: number;
  manager_id?: number;
  start_date?: string;
  end_date?: string;
  estimated_hours?: number;
  estimated_budget?: number;
  project_weight?: number;
}

export interface ProjectUpdate extends Partial<ProjectCreate> {
  reason?: string;
}

export interface DashboardSummary {
  total_projects: number;
  status_breakdown: Record<string, number>;
  overdue_projects: number;
  avg_progress: number;
  active_projects: number;
}

export interface MyProject {
  project_id: number;
  project_code: string | null;
  project_name: string;
  description: string | null;
  department: string;
  project_status: ProjectStatus;
  project_role: MemberRole | null;
  contribution_share: number;
  start_date: string | null;
  due_date: string | null;
  progress: number;
  project_health: 'OK' | 'AT_RISK' | 'OVERDUE' | 'COMPLETED' | string;
  assigned_tasks: number;
  doing_tasks: number;
  review_tasks: number;
  done_tasks: number;
  overdue_tasks: number;
}

export interface TeamWorkload {
  user_id: number;
  full_name: string;
  email: string;
  active_projects: number;
  assigned_tasks: number;
  doing_tasks: number;
  review_tasks: number;
  overdue_tasks: number;
  estimated_hours: number;
  actual_hours: number;
  workload_status: string;
}
